nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/FileWriters/XmlFileWriter.cs

```diff
@@ -0,0 +1,465 @@
+                        .ToArray();
+                    if (allPackageVersionElements.Length > 0)
+                    {
+                        // add a new `<PackageVersion>` element
+                        var newVersionElement = new XElement(XName.Get(PackageVersionElementName, projectDocument.Root.Name.NamespaceName),
+                            new XAttribute(IncludeAttributeName, requiredPackageVersion.Name),
+                            new XAttribute(VersionMetadataName, requiredVersion.ToString()));
+                        var lastPriorPackageVersionElement = allPackageVersionElements
+                            .TakeWhile(e => (e.Attribute(IncludeAttributeName)?.Value ?? string.Empty).Trim().CompareTo(requiredPackageVersion.Name) < 0)
+                            .LastOrDefault();
+                        if (lastPriorPackageVersionElement is not null)
+                        {
+                            _logger.Info($"Adding new `<{PackageVersionElementName}>` element for {requiredPackageVersion.Name} with version {requiredVersion}.");
+                            AddAfterSiblingElement(lastPriorPackageVersionElement, newVersionElement);
+                        }
+                        else
+                        {
+                            // no prior package versions; add to the front of the document
+                            _logger.Info($"Adding new `<{PackageVersionElementName}>` element for {requiredPackageVersion.Name} with version {requiredVersion} at the start of the document.");
+                            var packageVersionGroup = allPackageVersionElements.First().Parent!;
+                            var indent = GetIndentXTextFromElement(packageVersionGroup, extraIndentationToAdd: "  ");
+                            packageVersionGroup.AddFirst(indent, newVersionElement);
+                        }
+                    }
+                    else
+                    {
+                        // add a direct `Version` attribute
+                        newElement.SetAttributeValue(VersionMetadataName, requiredVersion.ToString());
+                    }
+                }
+            }
+            else
+            {
+                // found matching `<PackageReference>` elements to update
+                foreach (var packageReferenceElement in packageReferenceElements)
+                {
+                    // first check for matching `Version` attribute
+                    var versionAttribute = packageReferenceElement.Attributes().FirstOrDefault(a => a.Name.LocalName.Equals(VersionMetadataName, StringComparison.OrdinalIgnoreCase));
+                    if (versionAttribute is not null)
+                    {
+                        currentVersionString = versionAttribute.Value;
+                        updateVersionLocation = (version) => versionAttribute.Value = version;
+                        goto doVersionUpdate;
```

Comment:
I'm aware of "goto considered harmful" and "goto considered harmful considered harmful", but this is the first time I've ever seen `goto` in production code 😅